src/io/g15.rs

```diff
@@ -0,0 +1,251 @@
+    let mut player_ref = players.get_mut(idx).ok_or(G15BadIndex::IndexOutOfBounds)?;
+    player_ref.valid = Some(truth_value);
+    Ok(())
+}
+
+// m_iUserID[2] integer (68)    --match--> (player idx) (server id (as it appears in status))
+pub const REGEX_I_USERID: &str = r#"^m_iUserID\[(\d+)\]\s+integer\s+\((\d+)\)$"#;
+pub fn parse_userid(caps: Captures, players: &mut Vec<G15Player>) -> Result<(), G15BadIndex> {
+    let idx: usize = caps[1].parse().unwrap();
+    let userid = caps[2].to_string();
+    let mut player_ref = players.get_mut(idx).ok_or(G15BadIndex::IndexOutOfBounds)?;
+    player_ref.userid = Some(userid);
+    Ok(())
+}
+
+
+// The G15Player struct contains all the data per player searched.
+// Each of the elements may be Null, as the parsing could fail on a particular line, leaving the value undefined.
+// We still want the rest of the useful data though
+#[derive(Debug, Clone)]
+pub struct G15Player {
+    pub name: Option<String>,     // eg "Lilith"
+    pub ping: Option<u32>,        // eg 21
+    pub score: Option<u32>,       // eg 16
+    pub deaths: Option<u32>,      // eg 5
+    pub sid3: Option<String>,     // eg [U:1:111216987]
+    pub team: Option<u32>,        // eg 3
+    pub health: Option<u32>,      // eg 125
+    pub ammo: Option<u32>,        // eg 6
+    pub connected: Option<bool>,  // eg true
+    pub valid: Option<bool>,      // eg true
+    pub alive: Option<bool>,      // eg true
+    pub userid: Option<String>,   // eg "301"
+}
+impl G15Player {
+    pub fn new() -> G15Player {
+        G15Player {
+            name: None, 
+            ping: None, 
+            score: None, 
+            deaths: None, 
+            sid3: None, 
+            team: None, 
+            health: None, 
+            ammo: None,
+            alive: None,
+            connected: None,
+            userid: None,
+            valid: None,
+        }
+    }
+
+    
+}
+
+pub struct G15Parser {
+    patterns: Vec<RegMatch>,
+}
+impl G15Parser {
+    pub fn new() -> G15Parser {
+        G15Parser { patterns: vec![
+            RegMatch::new(REGEX_I_AMMO, parse_ammo), 
+            RegMatch::new(REGEX_SZ_NAME, parse_name),
+            RegMatch::new(REGEX_I_PING, parse_ping),
+            RegMatch::new(REGEX_I_SCORE, parse_score),
+            RegMatch::new(REGEX_I_DEATHS, parse_deaths),
+            RegMatch::new(REGEX_B_CONNECTED, parse_connected),
+            RegMatch::new(REGEX_I_TEAM, parse_team),
+            RegMatch::new(REGEX_B_ALIVE, parse_alive),
+            RegMatch::new(REGEX_I_HEALTH, parse_health),
+            RegMatch::new(REGEX_I_SID3, parse_sid3),
+            RegMatch::new(REGEX_B_VALID, parse_valid),
+            RegMatch::new(REGEX_I_USERID, parse_userid),
+            ] }
+    }
+    /*
+        Parse a g15_dumpplayer string via Regex search.
+        We only extract useful data here, so drop most data.
+    */
+    pub fn parse_g15(&self, g15_log: String) -> Vec<G15Player> {

```

Comment:
If you can, this should be changed to take `&str` instead of `String`

Reply:
because we want to borrow a string from the rcon stream, right?

Reply:
Well we don't know exactly what will happen at the call site of this function, but having to only pass a `&str` is much more flexible as the caller doesn't need to own the string either and it may avoid an extra string clone in many cases.